In [17]:
import pandas as pd
import numpy as np

df = pd.read_csv("kc_house_data.csv")
df = df.drop(columns=["id", "date", "zipcode"])

y = df["price"]             
X = df.drop(columns=["price"])

## 5.1:

In [22]:
# Computing mean, min, max, variance for each X feature
summary = X.agg(["mean", "min", "max", "var"]).T
summary

mean        min           max           var
bedrooms           3.370842     0.0000  3.300000e+01  8.650150e-01
bathrooms          2.114757     0.0000  8.000000e+00  5.931513e-01
sqft_living     2079.899736   290.0000  1.354000e+04  8.435337e+05
sqft_lot       15106.967566   520.0000  1.651359e+06  1.715659e+09
floors             1.494309     1.0000  3.500000e+00  2.915880e-01
waterfront         0.007542     0.0000  1.000000e+00  7.485226e-03
view               0.234303     0.0000  4.000000e+00  5.872426e-01
condition          3.409430     1.0000  5.000000e+00  4.234665e-01
grade              7.656873     1.0000  1.300000e+01  1.381703e+00
sqft_above      1788.390691   290.0000  9.410000e+03  6.857347e+05
sqft_basement    291.509045     0.0000  4.820000e+03  1.958727e+05
yr_built        1971.005136  1900.0000  2.015000e+03  8.627973e+02
yr_renovated      84.402258     0.0000  2.015000e+03  1.613462e+05
lat               47.560053    47.1559  4.777760e+01  1.919990e-02
long            -122.213896  -122.5190 -1.213150e+02  1.983262e-02
sqft_living15   1986.552492   399.0000  6.210000e+03  4.697612e+05
sqft_lot15     12768.455652   651.0000  8.712000e+05  7.455182e+08

In [24]:
lowest_mean_feature  = summary["mean"].idxmin()
highest_mean_feature = summary["mean"].idxmax()

lowest_mean_value  = summary.loc[lowest_mean_feature, "mean"]
highest_mean_value = summary.loc[highest_mean_feature, "mean"]

lowest_mean_feature, lowest_mean_value, highest_mean_feature, highest_mean_value


('long', -122.21389640494147, 'sqft_lot', 15106.967565816869)

In [25]:
lowest_var_feature  = summary["var"].idxmin()
highest_var_feature = summary["var"].idxmax()

lowest_var_value  = summary.loc[lowest_var_feature, "var"]
highest_var_value = summary.loc[highest_var_feature, "var"]

lowest_var_feature, lowest_var_value, highest_var_feature, highest_var_value


('waterfront', 0.007485225502689098, 'sqft_lot', 1715658774.1754541)

### Answer:

For the features considered, the feature with the lowest average value is longitude, with a mean of ~ −122.21, reflecting the negative longitude values in the dataset. The feature with the highest average value is sqft_lot, with a mean of ~ 15,106.97, indicating that lot sizes are large on average compared to other features. 

Waterfront has the lowest variance (~ 0.0075), which is expected since it is a binary feature indicating the presence of a waterfront. The feature with the highest variance is sqft_lot, with a variance of ~ 1.72 × 10⁹, showing substantial variability in lot sizes across the dataset.

## 5.2:

In [26]:
corr_with_price = X.corrwith(y).sort_values(ascending=False)
corr_with_price

sqft_living      0.702035
grade            0.667434
sqft_above       0.605567
sqft_living15    0.585379
bathrooms        0.525138
view             0.397293
sqft_basement    0.323816
bedrooms         0.308350
lat              0.307003
waterfront       0.266369
floors           0.256794
yr_renovated     0.126434
sqft_lot         0.089661
sqft_lot15       0.082447
yr_built         0.054012
condition        0.036362
long             0.021626
dtype: float64

### Answer:

The Pearson correlation coefficients show that all features are positively correlated with house price, meaning that increases in any of the considered features are associated with higher prices. 

The feature with the strongest positive correlation with price is sqft_living (correlation ≈ 0.70), indicating that larger living area is most strongly associated with higher house prices. 

Other features like grade, sqft_above, and bathrooms also show relatively strong positive correlations with price, while features like condition and longitude have weaker positive relationships. 

No features were found to have a negative correlation with the response variable.

## 5.3:

### Answer:

No, there are no negative correlations with price.